Redesign banners in organization settings.

shuklamaneesh23 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #34252.

**Screenshots and screen captures:**


| **Location**                                             | **Before**                                                                                                                                      | **After**                                                                                                                                       |
|----------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|
| ORGANIZATION PROFILE                                     | <img width="871" alt="Before 1" src="https://github.com/user-attachments/assets/9fef84e3-edfc-4a0f-8701-90dc3108f1d9" />                        | <img width="880" alt="After 1" src="https://github.com/user-attachments/assets/dfd59945-b9a8-461f-89c1-f5c0ee852095" />                        |
| Channel settings                                         | <img width="704" alt="Before 2" src="https://github.com/user-attachments/assets/e422cdd7-1723-41e8-bd83-224c4704d08b" />                        | <img width="705" alt="After 2" src="https://github.com/user-attachments/assets/b9b04201-64a5-4b14-b44a-5cf538a47893" />                        |
| AUTHENTICATION METHODS                                   | <img width="875" alt="Before 3" src="https://github.com/user-attachments/assets/17cfe11e-e634-4b99-b743-e86409d41b2e" />                        | <img width="869" alt="After 3" src="https://github.com/user-attachments/assets/4137980c-a891-49b3-9504-d27476a80805" />                        |
| DEFAULT USER SETTINGS                                    | <img width="873" alt="Before 4" src="https://github.com/user-attachments/assets/02772b42-31ca-4715-896e-7245f8e34f14" />                        | <img width="873" alt="After 4" src="https://github.com/user-attachments/assets/00fa4b00-8d50-44d3-8f48-2c1e87b93ccc" />                        |
| SETTINGS / BOTS                                          | <img width="875" alt="Before 5" src="https://github.com/user-attachments/assets/7d8175fc-8b2a-4e63-95f5-ad5d0f517de3" />                        | <img width="878" alt="After 5" src="https://github.com/user-attachments/assets/47e94362-2dc6-4ce4-a952-7d9a6686a5d2" />                        |
| Upgrade in org logo in new realm                         | <img width="827" alt="Before 6" src="https://github.com/user-attachments/assets/630703ef-1dbb-4e6e-941e-f9598c1cdfbe" />                        | <img width="800" alt="After 6" src="https://github.com/user-attachments/assets/78f4850c-e287-42dc-85ff-d9fe787e73a9" />                        |
| Upgrade in group settings in new realm                   | <img width="525" alt="Before 7" src="https://github.com/user-attachments/assets/c7c61b2b-6bda-4e87-a270-eaf2eee29e61" />                        | <img width="600" alt="After 7" src="https://github.com/user-attachments/assets/4f051139-8f70-4dfd-ab29-542200cadcfa" />                        |
| Upgrade or sponsor in group settings in a demo org       | <img width="683" alt="Before 9" src="https://github.com/user-attachments/assets/8a5f612b-3988-4942-bd44-1e940d0064c2" />                        | <img width="605" alt="Screenshot 2025-04-27 at 1 56 08 PM" src="https://github.com/user-attachments/assets/e4e905b8-ce94-4497-839f-e475dcf5dd91" />                      |
| Channel settings learn more | <img width="724" alt="Screenshot 2025-04-26 at 2 53 49 PM" src="https://github.com/user-attachments/assets/a74d25fc-d903-4684-89fa-4055dd14b0ba" /> | <img width="722" alt="Screenshot 2025-04-26 at 3 31 53 PM" src="https://github.com/user-attachments/assets/c7209122-1d9e-49f4-bf1d-48416f00115f" /> |
| User Group in new realm | <img width="722" alt="Screenshot 2025-04-26 at 2 53 37 PM" src="https://github.com/user-attachments/assets/b622ab23-85b0-4044-af3e-b847a12bf920" /> | <img width="721" alt="Screenshot 2025-04-26 at 2 47 32 PM" src="https://github.com/user-attachments/assets/894ffd84-a380-435a-9c84-4af3aa1e0333" /> |
| Upgrade banner for non-admin | <img width="372" alt="Screenshot 2025-04-26 at 2 55 16 PM" src="https://github.com/user-attachments/assets/0ffc3338-9002-40a5-a493-cebc90af4017" /> | <img width="327" alt="Screenshot 2025-04-26 at 2 48 09 PM" src="https://github.com/user-attachments/assets/58cf739e-2fd0-4cec-865e-12132f285100" /> |
| User Group in demo org | <img width="724" alt="Screenshot 2025-04-26 at 2 54 59 PM" src="https://github.com/user-attachments/assets/0075c5bf-f3e1-4042-bc79-80405def0eae" /> | <img width="721" alt="Screenshot 2025-04-26 at 2 48 46 PM" src="https://github.com/user-attachments/assets/7bf6480b-ce6f-453d-8808-73e44b5215c8" /> |



<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sayamsamal commented:
> Medium emphasis menas `"attention: quiet"`, so do low emphasis means `"attention: borderless"`

Yup, that is the correct relation between them, so you'd want the banner to look like the one you attached in the comments.

alya commented:
Testing the PR, the banner sizes and positioning look very hodge-podge.

Let's keep them left-aligned like everything else, and always on their own dedicated line. We should also set some reasonable min width for upgrade banners -- matching the inputs and dropdowns would look reasonable.

alya commented:
I'm thinking for non-admin users, we should make the upgrade banners neutral, as otherwise they draw more attention than necessary.

alya commented:
The mismatch between these banners doesn't look great, so I think we should update both on the same PR.

![Screenshot 2025-04-25 at 17 07 38](https://github.com/user-attachments/assets/25a24e9e-ae82-4660-a164-c54bd9594410)

alya commented:
I haven't done a fully comprehensive pass, but looks good when I poke around!

I also noticed that we forgot to do this banner:

![Screenshot 2025-04-29 at 23 29 38@2x](https://github.com/user-attachments/assets/a45338e8-6077-42a2-9642-9edabc696f47)

I'll add it to the issue description now. You can ping @sahil839 for review once you've done that one.

sayamsamal commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,136 @@
+import {$t} from "./i18n.ts";
+import * as settings_data from "./settings_data.ts";
+
+const UPGRADE_ACCESS_BANNER: banners.Banner = {
```
You can do an additional `import type {Banner} from "./banners.ts";` and avoid having to write `banners.Banner` again and again.

sayamsamal commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,136 @@
+        window.open("/help/introduction-to-channels", "_blank", "noopener,noreferrer");
+    });
+
+    $(() => {
+        initialize_upgrade_banners();
+    });
```
Is it necessary to run this line of code on DOM load given that we are calling the `set_up` method in the other modules when necessary?

shuklamaneesh23 replied:
Thanks for pointing this out. I see that it’s no longer necessary, since we’re already calling the set_up method in the modules as needed.

sayamsamal commented:
@shuklamaneesh23 I had a quick pass through your code, and the general direction looks good to me. @sahil839 will give you a proper review, after you have done the changes.

shuklamaneesh23 commented:
@sayamsamal , could you please check the banner responsiveness at the narrow screens?

sahil839 commented on web/src/settings_org.ts:
```diff
@@ -1320,6 +1321,7 @@ export function build_page(): void {
     meta.loaded = true;
 
     loading.make_indicator($("#admin_page_auth_methods_loading_indicator"));
+    settings_banner.set_up();
```
Calling `set_up` here, and from groups and stream settings related file as well, means that "click" handlers are being set up multiple times and that can result in multiple tabs being opened for the same link.

sahil839 commented on web/templates/settings/notification_settings.hbs:
```diff
@@ -157,11 +157,7 @@
-                Mobile push notifications are not enabled on this server. <z-link>Learn more</z-link>
-                {{#*inline "z-link"}}<a href="/help/mobile-notifications#enabling-push-notifications-for-self-hosted-servers" target="_blank" rel="noopener noreferrer">{{> @partial-block}}</a>{{/inline}}
-            {{/tr}}
+        <div class="mobile-push-notifications-banner-container banner-wrapper">
```
There is no space between this banner and the checkbox setting.

sahil839 replied:
This banner is only visible once I open organization settings UI atleast once, but not visible when I open personal settings UI directly.

sahil839 commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,164 @@
+    const $org_upgrade_container = $(".organization-upgrade-banners-container");
+    if ($org_upgrade_container.length > 0 && has_billing_access) {
+        banners.open(UPGRADE_ACCESS_BANNER, $org_upgrade_container);
+    } else {
```
Upgrade banner is not visible for message retention setting in stream settings UI.

sahil839 reviewed (commented):
Posted some comments.

kuv2707 commented:
Looks like this PR is well beyond `buddy review`.

sahil839 commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,170 @@
+        banners.open(banner, $upgrade_or_sponsership_container);
+    }
+
+    $(document).on("click", ".request-upgrade", (e) => {
```
Having these click handlers here results in opening the link in two new tabs for cases where this function is called twice. These handlers should be initialized only once.

sahil839 commented on web/styles/app_components.css:
```diff
@@ -542,10 +542,32 @@ input.settings_text_input {
+}
+
+#organization-profile {
+    .banner {
```
Here also a more specific selector can be used.

sahil839 commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,170 @@
+export function initialize_upgrade_banners(): void {
+    const has_billing_access = settings_data.user_has_billing_access();
+    const $org_upgrade_container = $(".organization-upgrade-banners-container");
+    if ($org_upgrade_container.length > 0) {
```
Can the check for `is_business_type_org` be here itself instead of having that in `upgrade_tip_widget.hbs`? That would make it easy for someone reading the code as all the logic to decide which banner to show will be in one place and only one container element will be needed.

sahil839 commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,170 @@
+    intent: "neutral",
+    label: $t({defaultMessage: "Available on Zulip Cloud Standard."}),
+    buttons: [],
+    close_button: false,
```
This and `UPGRADE_BANNER` are exactly same, maybe we can use the same object by having a better name.

sahil839 reviewed (commented):
Posted some comments.

shuklamaneesh23 commented:
Updates:

- Added check for `is_business_type_org` in the `settings_banner.ts` as instructed.
- Used custom classes for adding `margin-bottom` to several banners.
- Renamed `UPGRADE_BANNER` to `UPGRADE_TO_ZULIP_CLOUD_BANNER` so that it can be reused at other places.
- Kept the event listeners for `request-upgrade` and `request-sponsorship` in `web/src/click_handlers.ts` so that they get triggered only once.

sahil839 commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,163 @@
+
+    banners.open(MOBILE_PUSH_NOTIFICATION_BANNER, $container);
+
+    $container.on("click", ".mobile-push-notification-info", (e) => {
```
The link is opened twice if I click on the button after updating one of the settings in the table.

sahil839 reviewed (commented):
Posted few comments. Also, the banner for message retention setting needs to be fixed -
![Screenshot from 2025-06-09 15-30-15](https://github.com/user-attachments/assets/caaf5462-b004-48db-bf6e-9eb216f7a805)

shuklamaneesh23 commented:
> Posted few comments. Also, the banner for message retention setting needs to be fixed

@sahil839, addressed all the points.
The banner for message retention settings is as per the css defined for `banner-wrapper` when `<44em`.
For now, i've removed the `banner-wrapper` class from its parent to make its alignment as you want and have also posted question regarding this is desing thread. [CZO Link](https://chat.zulip.org/#narrow/channel/6-frontend/topic/Correct.20banner.20behaviour.20at.20narrower.20widths.2E/with/2190078)

sahil839 commented on web/src/settings_notifications.ts:
```diff
@@ -86,6 +87,7 @@ function rerender_ui(): void {
         $unmatched_streams_table.css("display", "table-row-group");
     }
     update_desktop_notification_banner();
+    settings_banner.set_up_mobile_push_banner();
```
I think there is no need to call this here, since showing the banner depends on `realm.realm_push_notifications_enabled` field and we do not live-update on changing that field. This can also be called in set_up.

sahil839 replied:
Also, the banner is not shown in "Default user settings" panel if I do not open the "Notifications" panel at least once.

sahil839 commented on web/src/settings_notifications.ts:
```diff
@@ -407,6 +409,15 @@ export function set_up(settings_panel: SettingsPanel): void {
     set_enable_marketing_emails_visibility();
     rerender_ui();
+
+    $container.on("click", ".mobile-push-notification-info", (e) => {
```
Clicking on "Learn more" does not work for "Default user settings" panel. And I guess click handlers for desktop notifiction banner do not work as well, which can be fixed in a separate PR.

shuklamaneesh23 replied:
Moved the event listener before the `return` is being called up in the `set_up` function and hence it's now working.

sahil839 reviewed (commented):
Posted a couple of comments.

shuklamaneesh23 commented:
Hi @sahil, please have a look now.
Aloso, for the bug here in `desktop notifiction banner`, i'll open a separate PR and ping you in it.
Thanks!

sahil839 commented:
Looks good to me. Not sure if we want to wait to merge this till the banner in stream settings gets fixed, as the button is on next line even on normal screen size and font size.

sahil839 commented:
The new commit looks good.

timabbott commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,154 @@
+        banner = has_billing_access ? UPGRADE_OR_SPONSORSHIP_BANNER : AVAILABLE_ON_STANDARD;
+    }
+    banners.open(banner, $upgrade_container);
+}
```
Is it safe to call `banners.open` on the same container element several times? This code path seems like it'll often do that.

shuklamaneesh23 replied:
Change `upgrade_container` to `upgrade_banner_containers`.

alya commented on web/src/attachments_ui.ts:
```diff
@@ -73,10 +74,30 @@ function set_upload_space_stats(): void {
+        buttons = [
+            ...buttons,
+            {
+                label: $t({defaultMessage: "Upgrade for more space."}),
```
Let's move this sentence into the text part of the banner, and then the button will just say "Upgrade". Our current pattern is to aim for short button labels.

shuklamaneesh23 replied:
Done.

sayamsamal commented on web/templates/settings/bot_settings_tip.hbs:
```diff
@@ -1,7 +1,19 @@
 {{#unless can_create_any_bots}}
     {{#if can_create_incoming_webhooks}}
-        <div class='tip'>{{t "You can create bots that can only send messages." }}</div>
+        {{> ../components/banner
```
The banners should always be wrapped inside a `banner-wrapper` parent for the container queries to work.

sayamsamal replied:
FYI, You can add the `banner-wrapper` class to the pre-existing parent, and don't always have to introduce a new wrapper div.

timabbott replied:
@shuklamaneesh23 have you resolved this feedback?

sayamsamal commented on web/src/stream_settings_ui.ts:
```diff
@@ -51,6 +52,21 @@ import type {StreamSubscription} from "./sub_store.ts";
 import * as util from "./util.ts";
 import * as views_util from "./views_util.ts";
 
+const STREAM_INFO_BANNER: Banner = {
```
This commit should be split into multiple ones targeting specific banners — Upgrade banner, Group Info banner, Stream Info banner, and Mobile Push Notifications banner. A common indicator as to when you should be splitting the commit would be when you start using compound sentences in the commit message with multiple conjunctions.

alya commented:
1. Let's change all the banners to be as wide as needed, rather than taking the entire width of the window. I think that will look better throughout, and especially whenever the banner is not at the top.
<img width="855" height="259" alt="Screenshot 2025-07-29 at 10 54 53" src="https://github.com/user-attachments/assets/b2101684-918a-4997-8bcb-9efe42198095" />
<img width="867" height="145" alt="Screenshot 2025-07-29 at 10 58 02" src="https://github.com/user-attachments/assets/d889174a-7120-4ea3-88d0-d8eb61bbdc0a" />

2. Layout problem:
<img width="565" height="264" alt="Screenshot 2025-07-29 at 10 55 32" src="https://github.com/user-attachments/assets/808f977a-0395-4be9-bc00-59eda3617343" />

3. Let's remove the permissions banners from Settings / Users / Users and Settings / Users / Deactivated. The current text doesn't make sense, and I don't think those panels need a banner.
<img width="886" height="173" alt="Screenshot 2025-07-29 at 10 59 14" src="https://github.com/user-attachments/assets/30e8a474-7089-4648-a2ea-0de2896aee67" />

4. This banner needs to be updated to the new design (blue, as usual).
<img width="869" height="268" alt="Screenshot 2025-07-29 at 10 59 38" src="https://github.com/user-attachments/assets/4d1cb8e6-e0e8-4916-8790-898e1d314eaf" />

5. This banner should not appear in org settings, only in personal settings. I feel like I've seen a comment about this before?
<img width="856" height="402" alt="Screenshot 2025-07-29 at 11 00 30" src="https://github.com/user-attachments/assets/453a9684-1cb9-4164-817d-f41494094cc4" />

6. The banner version with an upgrade button is incorrectly being shown to a non-admin user.
<img width="876" height="238" alt="Screenshot 2025-07-29 at 11 00 59" src="https://github.com/user-attachments/assets/fd040d2e-0fb4-4aee-967d-21eef5726827" />

I haven't really tested responsiveness and whether things work fine in other languages in this pass. I can do that once the feedback above has been addressed.

alya commented:
> Should I show them conditionally?

Yes. It's not useful to take a user to the page on the left. ;)

timabbott commented on web/src/settings_banner.ts:
```diff
@@ -0,0 +1,85 @@
+            attention: "borderless",
+        },
+    ],
+    custom_classes: "organization-upgrade-banner",
```
I feel like this code could be refactored; it'd be a lot cleaner to just define `upgrade_buttons`, the only part that varies, based on the parameters, without repeating the rest of this definition.

timabbott commented:
Merged, thanks @shuklamaneesh23 and @sayamsamal!

timabbott commented:
(Please follow up on the two items above, which likely have no functional impact)
